sync2/sqlstore/dbseq.go

```diff
@@ -0,0 +1,237 @@
+func idsFromTable(
+	db sql.Executor,
+	sts *SyncedTableSnapshot,
+	from rangesync.KeyBytes,
+	ts int64,
+	chunkSize int,
+	maxChunkSize int,
+	lru *lru,
+) rangesync.SeqResult {
+	if from == nil {
+		panic("BUG: makeDBIterator: nil from")
+	}
+	if maxChunkSize <= 0 {
+		panic("BUG: makeDBIterator: chunkSize must be > 0")
+	}
+	if chunkSize <= 0 {
+		chunkSize = 1
+	} else if chunkSize > maxChunkSize {
+		chunkSize = maxChunkSize
+	}
+	var err error
+	return rangesync.SeqResult{
+		Seq: func(yield func(k rangesync.KeyBytes) bool) {
+			s := &dbSeq{
+				db:           db,
+				from:         from.Clone(),
+				sts:          sts,
+				chunkSize:    chunkSize,
+				ts:           ts,
+				maxChunkSize: maxChunkSize,
+				keyLen:       len(from),
+				chunk:        make([]rangesync.KeyBytes, maxChunkSize),
+				singleChunk:  false,
+				cache:        lru,
+			}
+			if err = s.load(); err != nil {
+				return
+			}
+			err = s.iterate(yield)
+		},
+		Error: func() error {
+			return err
+		},
+	}
+}
+
+// loadCached loads a chunk of IDs from the LRU cache,
+// if possible.
+func (s *dbSeq) loadCached(key dbIDKey) (bool, int) {
+	if s.cache == nil {
+		return false, 0
+	}
+	chunk, ok := s.cache.Get(key)
+	if !ok {
+		return false, 0
+	}
+
+	for n, id := range s.chunk[:len(chunk)] {
+		if id == nil {
+			id = make([]byte, s.keyLen)
+			s.chunk[n] = id
+		}
+		copy(id, chunk[n])
+	}
+	return true, len(chunk)
+}
+
+// load makes sure the current chunk is loaded.
+func (s *dbSeq) load() error {
+	s.pos = 0
+	if s.singleChunk {
+		// we have a single-chunk DB sequence, don't need to reload,
+		// just wrap around
+		return nil
+	}
+
+	n := 0
+	// make sure the chunk is large enough
+	if cap(s.chunk) < s.chunkSize {
+		s.chunk = make([]rangesync.KeyBytes, s.chunkSize)
+	} else {
+		// if the chunk size was reduced due to a short chunk before wraparound, we need
+		// to extend it back
+		s.chunk = s.chunk[:s.chunkSize]
+	}
+	key := dbIDKey{string(s.from), s.chunkSize}
+	var ierr, err error
+	found, n := s.loadCached(key)
+	if !found {
+		dec := func(stmt *sql.Statement) bool {
+			if n >= len(s.chunk) {
+				ierr = errors.New("too many rows")
+				return false
+			}
+			// we reuse existing slices when possible for retrieving new IDs
+			id := s.chunk[n]
+			if id == nil {
+				id = make([]byte, s.keyLen)
+				s.chunk[n] = id
+			}
+			stmt.ColumnBytes(0, id)
+			n++
+			return true
+		}
+		if s.ts <= 0 {
+			err = s.sts.LoadRange(s.db, s.from, s.chunkSize, dec)
+		} else {
+			err = s.sts.LoadRecent(s.db, s.from, s.chunkSize, s.ts, dec)
+		}
+		if err == nil && ierr == nil && s.cache != nil {
+			cached := make([]rangesync.KeyBytes, n)
+			for n, id := range s.chunk[:n] {
+				cached[n] = slices.Clone(id)
+			}
+			s.cache.Add(key, cached)
+		}
+	}
+	fromZero := s.from.IsZero()
+	s.chunkSize = min(s.chunkSize*2, s.maxChunkSize)
+	switch {
+	case err != nil || ierr != nil:
+		return errors.Join(ierr, err)
```

Comment:
Returning both errors doesn't make sense to me, I think `ierr` should take precedence since the db request was interrupted because of it and `err` isn't interesting in this case, i.e.

```
case ierr != nil:
    return ierr
case err != nil:
    return err
case n == 0:
    // unchanged
...
```

Reply:
Fixed, thanks